### DISCUSSION TITLE
Provide a "last-modified" header for https://choria.io/RELEASE-GPG-KEY

### DISCUSSION THREAD

**Discussion by dploeger** [choria-io/general#47] [RFC]
After complying with the new APT keyring features, the keyring will be downloaded from https://choria.io/RELEASE-GPG-KEY every single time Puppet runs.

Would it be possible to provide a "last-modified" header on it so it isn't updated every time?

**Comment by ripienaar** [comment]
Hmm this is hosted on netlify iirc not sure I have control over that. 

But it’s certainly not good that this is being downloaded each time. Any ideas @smortex ?

**Reply by smortex** [reply]
The `file` resource has a [replace](https://www.puppet.com/docs/puppet/8/types/file.html#file-attribute-replace) attribute that tells Puppet to ignore the file content if the file is there.  Keys are not supposed to change often, so it is somewhat acceptable, but when they do change, I bet we will shoot in our own foot…

I quickly tested the `checksum` / `checksum_value` parameters to hard-code the checksum of the file in the Puppet manifest.  According to the `--debug` output, it only download the file if the checksum is not the expected one.  That would also limit the impact of a compromised file server where the key is replaced and rogue packages are delivered.


```puppet
file { '/tmp/KEY':
  ensure         => file,
  checksum       => 'md5', # sha256 or sha512 is maybe preferable
  checksum_value => '422b78010c402b95b502bca70f4e61d7',
  source         => 'https://choria.io/RELEASE-GPG-KEY',
}
```

Does this seems reasonable?

**Reply by dploeger** [reply]
The problem is that this would need to be implemented in the apt module. An alternative could be to use the archive module to download the keyring and provide that as the source. Archive has more possibilities.

But fixing that problem on the source is the best option IMHO - IF that's actually possible for you of course.

**Reply by ripienaar** [reply]
I could host the file on `apt.[eu|uk|us].choria.io` but there's no round robin address that would make that reliable so pointing to one of those alone wouldnt be good - apt itself uses mirrorlists.choria.io which isnt https cos old debian doesnt like it.  

Though old debian is now dead so we could switch that over but would require a bunch of testing

**Reply by smortex** [reply]
After updating the module, I see that not only the file is downloaded on each run, but puppet also report it as changed (I failed to understand this from the previous discussion).

So this (static) website is served by a service provider that does not handle file date properly?  That's hard to believe…

**Reply by ripienaar** [reply]
Yeah their CDN records time that the cache node got  the file not when the file was changed - and then doesn’t add this header

**Comment by smortex** [comment]
I was thinking about pointing to the "source" file (raw file in the GitHub repo of the website) but GitHub does not provide such header too (I guess it is costly to extract it from git with low added value :shrug:).  We can workaround this by embedding the key in the module: not great, but should be better that downloading the key every half hour on all nodes, and this avoid an HTTP request per run to check if a new key is available.  Of course, updating the key becomes more painful (update the key, update the module to use the new key, ship a new version of the module, update the version of the module used)…  What do you think?

**Comment by ripienaar** [comment]
Does the module support multiple urls to try if one fails?

**Reply by smortex** [reply]
The apt module does not currently support that. It is quite a niche need, but I guess we can contribute support for this in the puppetlabs-apt module and bump the minimum version accordingly when released.  But what would be the plan?  Have a "reliable" URL first where a change is not registered on every run, and if the key has to be changed, make that URL non-fetchable to fall-back on the non-reliable one?

**Reply by ripienaar** [reply]
The files are living stable on https://apt.eu.choria.io/RELEASE-GPG-KEY https://apt.uk.choria.io/RELEASE-GPG-KEY and https://apt.us.choria.io/RELEASE-GPG-KEY. What I dont have is some kind of stable single hostname thats going to handle a case if one of these are unavailable. 

Without the multi url support we cant list multiples this would have helped.  

So I probably need to host these files elsewhere where they are reliable and served by https. Mirror lists would work but its not HTTPS due to debian bugs.

**Reply by dploeger** [reply]
Maybe make the URL configurable then. At least this would put an end to the current situation. I'd rather live with one of the host not being accessible and switching the URLs than the constant stream of Puppet reports without modifications. 😄

**Reply by ripienaar** [reply]
Soonish will have one of the mirrors spanning 2 machines then I will put the gpg files there

**Reply by ripienaar** [reply]
Well I gave it a good go but getting 2 servers behind a LB at hetzner have beaten me, for lets encrypt they need the entire domain in their dns servers and I am not doing that.

So the remaining option is to put them on mirrorlists.choria.io - but this is a HTTP server not HTTPS.  It is though a RR between 3 machines so that addresses the availability concern.

Would it be acceptible to fetch this over HTTP not HTTPS?

**Reply by smortex** [reply]
> Would it be acceptible to fetch this over HTTP not HTTPS?

/set mode=paranoid

Doesn't it make it possible to distribute a rogue version from a MITM position (serve a rogue key (over http), a rogue list of URL (over http) to rogue repos where rogue packages can be fetched and successfully verified against the rogue key)?

Maybe not trivial, but that sound reasonably doable.

Having a copy of the key in the module feels more robust. Other than when the key is rotated, are there other situation this is a bit more annoying that fetching the key from the internet?

**Reply by ripienaar** [reply]
Indeed I had same concerns with that. 

Tbh looking at the repo logs the keys are least of the problems. Debian machines fetching the InRelease file all the time is worse hehe

**Reply by ripienaar** [reply]
Thinking I can host this in a github repo - even github static hosting doesnt provide these headers:

```
HTTP/2 200 
cache-control: max-age=300
content-security-policy: default-src 'none'; style-src 'unsafe-inline'; sandbox
content-type: image/png
etag: W/"40c97a40fc91c140af1fd5381148cc3d76bf8b0253a9ed3fb4661932e1751b2a"
strict-transport-security: max-age=31536000
x-content-type-options: nosniff
x-frame-options: deny
x-xss-protection: 1; mode=block
x-github-request-id: 7C6C:0D8B:286EFA1:2A6073D:65E5E297
accept-ranges: bytes
date: Mon, 04 Mar 2024 15:04:23 GMT
via: 1.1 varnish
x-served-by: cache-mrs10548-MRS
x-cache: HIT
x-cache-hits: 1
x-timer: S1709564664.978446,VS0,VE1
vary: Authorization,Accept-Encoding,Origin
access-control-allow-origin: *
cross-origin-resource-policy: cross-origin
x-fastly-request-id: 53e289633178aa8ddd0ead1420911e2ce54bdb72
expires: Mon, 04 Mar 2024 15:09:23 GMT
source-age: 94
content-length: 20177
X-Firefox-Spdy: h2
```

Starting to think its just not something tools like Puppet should be relying on anymore :(

**Reply by ripienaar** [reply]
However - while netlify and github do not return the `Last-Modified` header they do correctly handle the `If-Modified-Since` header:

```
$ curl -vO https://choria.io/RELEASE-GPG-KEY -H "If-Modified-Since:Wed, 01 Jun 2024 08:00:00 GMT"
....
HTTP/2 304
cache-control: public,max-age=0,must-revalidate
cache-status: "Netlify Edge"; hit
date: Mon, 04 Mar 2024 15:11:33 GMT
etag: "f033695a93883c893214fb4c23807fba-ssl"
server: Netlify
strict-transport-security: max-age=31536000
x-nf-request-id: 01HR50A9KMJ1YF72BTJ6GXKCA1
```

So this appear to work fine, why does puppet need the modified header, surely it uses file mtime on the node?

**Reply by dploeger** [reply]
Well, it's just how they do it: https://github.com/puppetlabs/puppet/blob/d068941191ba289825b44343fec1a3d3f1173022/lib/puppet/type/file/source.rb#L56

With Last-Modified actually only being a fallback for missing X-Checksum-* headers.

**Reply by ripienaar** [reply]
Not sure what to say, it's just not how the biggest hosting offerings globally function :)

So I guess we should make the url to the key configurable, but its about as bad as embedding it in the module, both means everyone will have a bad time if I ever have to update that key

**Reply by dploeger** [reply]
As an alternative it would be possible to use the Archive module to download the key file and use what comes out of that as the source instead of just providing the url.

**Reply by ripienaar** [reply]
Can you see if https://static.choria.io/RELEASE-GPG-KEY will work for you?

**Reply by dploeger** [reply]
It does, thanks a lot. I'm preparing a PR for it.

**Reply by dploeger** [reply]
See https://github.com/choria-io/puppet-choria/pull/354
